telethon/client/updates.py

```diff
@@ -415,18 +415,60 @@ async def _update_loop(self: 'TelegramClient'):
                 except GapError:
                     continue  # get(_channel)_difference will start returning requests
 
-                updates_to_dispatch.extend(self._preprocess_updates(processed, users, chats))
+                updates_to_dispatch.extend(await self._preprocess_updates(processed, users, chats))
         except asyncio.CancelledError:
             pass
         except Exception as e:
             self._log[__name__].exception('Fatal error handling updates (this is a bug in Telethon, please report it)')
             self._updates_error = e
             await self.disconnect()
 
-    def _preprocess_updates(self, updates, users, chats):
-        self._mb_entity_cache.extend(users, chats)
-        entities = {utils.get_peer_id(x): x
-                    for x in itertools.chain(users, chats)}
+    async def _fetch_peers(self, users: list, chats: list):
+        result_users = {}
+        result_chats = {}
+        is_min = False
+        for peer in itertools.chain(users, chats):
+            if getattr(peer, 'min', False):
+                peer_id = peer.id
+                if not self._mb_entity_cache.get(peer_id) and peer.id != 136817688:
+                    is_min = True
+            if isinstance(peer, types.User):
+                result_users.update({utils.get_peer_id(peer): peer})
+            elif isinstance(peer, (types.Chat, types.Channel)):
+                result_chats.update({utils.get_peer_id(peer): peer})
+
+        return is_min, result_users, result_chats
+
+    async def _preprocess_updates(self, updates, users, chats):
+        is_min, users, chats = await self._fetch_peers(users, chats)
+        for update in updates:
+            if isinstance(update, (
+                    types.UpdateChannelTooLong,
+                    types.UpdateNewChannelMessage,
+                    types.UpdateEditChannelMessage,
+                    types.UpdateReadChannelInbox,
+                    types.UpdateReadChannelOutbox,
+            )):
+                if is_min and update.pts:
+                        get_diff = functions.updates.GetChannelDifferenceRequest(
+                            force=False,
+                            channel=await self.get_input_entity(update.message.peer_id),
```

Comment:
How does this work if the user is not in cache?

If `await self.get_input_entity` actually returns a working hash, that can be used to update the `_mb_entity_cache`, and there wouldn't be a need to make network calls.

Reply:
await self.get_input_entity returns a working hash, but channel message can contain more information than one channel or user, so I think it's better after all through GetChannelDifference. Perhaps this is not so

Reply:
If we consider this code as part of solving the problem when state is written and entity is not written, then there is probably no point in running GetChannelDifferenceRequest, because it will be enough to get hash through get_input_entity

Reply:
I revised my mistake (which I had previously sent to the chat) and got confused. After all, it was get_input_entity who gave an error that he could not find an entity with the necessary id when starting the script and catch_up=True. It's strange that I didn't get any testing errors. Perhaps I should check this code again

Reply:
I figured it out. When catch_up=True starts session.get_input_entity and when there is no entity in the cache, the script crashes with an error. I used the client.users.get_input_entity, which, as far as I understand, in case of absence in the cache, makes a request to the telegram api. What shall I do? We can run client.users.get_input_entity in the _preprocess_updates, we can run GetChannelDifference to get all the channels and all the users that are in this update, we can run client.users.get_input_entity when catch_up=True, but then there is a great risk of colliding with api timeouts

Reply:
> client.users.get_input_entity, which, as far as I understand, in case of absence in the cache, makes a request to the telegram api

This is only true when the input parameter passed to `get_input_entity` is a string (such as usernames or phone numbers), which, if not in cache, can make a request to fetch it. But if you pass an `id`, the library *must* have the corresponding `access_hash` or there is *no way* to fetch it otherwise. If there was a method to fetch `access_hash` from `id`, then `accecss_hash` would be useless (since they could always be fetched). `access_hash` is proof that the client has interacted with that `id` in some way before.

> We can run client.users.get_input_entity in the _preprocess_updates

If that works, then that's a fine solution. The `access_hash` would be loaded from the cache in the `.session` file. There would be no additional requests.

> we can run GetChannelDifference to get all the channels and all the users that are in this update

This is also an acceptable solution. But if the problem is that the *channel* itself is not in cache, I would expect it to not be possible to call `GetChannelDifference` at all since it *needs* an `access_hash`. I just don't know if the `min` one is enough, but if it is, that would solve our problems ([`utils.get_input_peer`](https://github.com/LonamiWebs/Telethon/blob/68ea208b4378263d2e04abd8357fe957770ec620/telethon/utils.py#L142) with `check_hash=True` would be a good way to turn the `Channel.min` into the input type, bypassing the default check).

> we can run client.users.get_input_entity when catch_up=True

Again, if that works, that's fine. Using the ID there won't make requests.

Reply:
I added even more variables and stopping points for debugging and realized that my code did not work as intended. Nothing works. I don't know how I managed to avoid the original problem and why entities were filled in the end, but when i start debug, i got then you are right, self.get_input_entity (update.message.peer_id) often none and GetChannelDifference do not work in most situations. I'm sorry. It turns out that this pr is meaningless and can be rejected. We need to look for other ideas for filling in entities

Reply:
Sorry to spend your time. I don't know how it happened